src/predictor/gpu_predictor.cu

```diff
@@ -268,28 +268,30 @@ class GPUPredictor : public xgboost::Predictor {
       max_shared_memory_bytes_ = dh::MaxSharedMemory(this->device_);
      }
     void PredictInternal
-    (const SparsePage& batch, const MetaInfo& info,
+    (const SparsePage& batch, bool is_first_batch, const MetaInfo& info,
      HostDeviceVector<bst_float>* predictions,
      const gbm::GBTreeModel& model,
      const thrust::host_vector<size_t>& h_tree_segments,
      const thrust::host_vector<DevicePredictionNode>& h_nodes,
      size_t tree_begin, size_t tree_end) {
       if (predictions->DeviceSize(device_) == 0) { return; }
       dh::safe_cuda(cudaSetDevice(device_));
-      nodes_.resize(h_nodes.size());
-      dh::safe_cuda(cudaMemcpyAsync(dh::Raw(nodes_), h_nodes.data(),
-                                    sizeof(DevicePredictionNode) * h_nodes.size(),
-                                    cudaMemcpyHostToDevice));
-      tree_segments_.resize(h_tree_segments.size());
-
-      dh::safe_cuda(cudaMemcpyAsync(dh::Raw(tree_segments_), h_tree_segments.data(),
-                                    sizeof(size_t) * h_tree_segments.size(),
-                                    cudaMemcpyHostToDevice));
-      tree_group_.resize(model.tree_info.size());
-
-      dh::safe_cuda(cudaMemcpyAsync(dh::Raw(tree_group_), model.tree_info.data(),
-                                    sizeof(int) * model.tree_info.size(),
-                                    cudaMemcpyHostToDevice));
+      if (is_first_batch) {
+        nodes_.resize(h_nodes.size());
+        dh::safe_cuda(cudaMemcpyAsync(dh::Raw(nodes_), h_nodes.data(),
+                                      sizeof(DevicePredictionNode) * h_nodes.size(),
+                                      cudaMemcpyHostToDevice));
+
+        tree_segments_.resize(h_tree_segments.size());
+        dh::safe_cuda(cudaMemcpyAsync(dh::Raw(tree_segments_), h_tree_segments.data(),
+                                      sizeof(size_t) * h_tree_segments.size(),
+                                      cudaMemcpyHostToDevice));
+
+        tree_group_.resize(model.tree_info.size());
+        dh::safe_cuda(cudaMemcpyAsync(dh::Raw(tree_group_), model.tree_info.data(),
+                                      sizeof(int) * model.tree_info.size(),
+                                      cudaMemcpyHostToDevice));
+      }
```

Comment:
This code would look better as a separate method. In my opinion, this looks more logical, and reduces the number of function parameters.

Reply:
Done.